src/server/game/Anticheat/AnticheatMgr.cpp

```diff
@@ -0,0 +1,482 @@
+{
+    uint32 key = player->GetGUID().GetCounter();
+    if (AnticheatData* data = GetPlayerData(key))
+    {
+        if (data->GetLastOpcode() == MSG_MOVE_JUMP && opcode == MSG_MOVE_JUMP)
+        {
+            TC_LOG_TRACE("entities.player.character", "AnticheatMgr:: Jump-Hack detected player GUID (low) %u", key);
+            return true;
+        }
+    }
+
+    return false;
+}
+
+bool TeleportPlaneCheck::OnCheck(Player* player, MovementInfo const& movementInfo, uint16 /*opcode = 0*/) const
+{
+    uint32 key = player->GetGUID().GetCounter();
+    if (AnticheatData* playerData = GetPlayerData(key))
+        if (playerData->GetLastMovementInfo().pos.GetPositionZ() != 0 ||
+            movementInfo.pos.GetPositionZ() != 0)
+            return false;
+
+    if (movementInfo.HasMovementFlag(MOVEMENTFLAG_FALLING))
+        return false;
+
+    //DEAD_FALLING was deprecated
+    //if (player->getDeathState() == DEAD_FALLING)
+    //    return;
+    float x, y, z;
+    player->GetPosition(x, y, z);
+    float ground_Z = player->GetMap()->GetHeight(x, y, z);
+    float z_diff = fabs(ground_Z - z);
+
+    // we are not really walking there
+    if (z_diff > 1.0f)
+    {
+        TC_LOG_TRACE("entities.player.character", "AnticheatMgr:: Teleport To Plane - Hack detected player GUID (low) %u", key);
+        return true;
+    }
+
+    return false;
+}
+
+// basic detection
+bool ClimbHackCheck::OnCheck(Player* player, MovementInfo const& movementInfo, uint16 opcode /*= 0*/) const
+{
+    uint32 key = player->GetGUID().GetCounter();
+    if (AnticheatData* playerData = GetPlayerData(key))
+        if (opcode != MSG_MOVE_HEARTBEAT || playerData->GetLastOpcode() != MSG_MOVE_HEARTBEAT)
+            return false;
+
+    // in this case we don't care if they are "legal" flags, they are handled in another parts of the Anticheat Manager.
+    if (player->IsInWater() ||
+        player->IsFlying() ||
+        player->IsFalling())
+        return false;
+
+    Position playerPos = player->GetPosition();
+
+    float deltaZ = fabsf(playerPos.GetPositionZ() - movementInfo.pos.GetPositionZ());
+    float deltaXY = movementInfo.pos.GetExactDist2d(&playerPos);
+
+    // Prevent divide by 0
+    if (G3D::fuzzyEq(deltaXY, 0.f))
+        return false;
+
+    float angle = Position::NormalizeOrientation(tangent(deltaZ / deltaXY));
+
+    if (angle > CLIMB_ANGLE)
+    {
+        TC_LOG_TRACE("entities.player.character", "AnticheatMgr:: Climb-Hack detected player GUID (low) %u", key);
+        return true;
+    }
+
+    return false;
+}
+
+void AnticheatMgr::StartHackDetection(Player* player, MovementInfo const& movementInfo, uint16 opcode)
+{
+    if (!sWorld->getBoolConfig(CONFIG_ANTICHEAT_ENABLE))
+        return;
+
+    // Disable anticheat for Game Masters
+    if (player->CanBeGameMaster())
+        return;
+
+    ObjectGuid::LowType lowGuid = player->GetGUID().GetCounter();
+    AnticheatData& plData = m_playerMap[lowGuid];
```

Comment:
Looks like it is safe to me entry in m_playerMap is created on login (threadsafe context) and destroyed on logout (threadsafe context)
Modifying a single map value is not unsafe for m_playerMap itself but can be unsafe for the value - however this is not a problem either because only one thread can process movement packets for one player.